storm-server/src/main/java/org/apache/storm/scheduler/resource/ResourceAwareScheduler.java

```diff
@@ -247,13 +234,169 @@ private void scheduleTopology(TopologyDetails td, Cluster cluster, final User to
                 }
             } catch (Exception ex) {
                 markFailedTopology(topologySubmitter, cluster, td,
-                                   "Internal Error - Exception thrown when scheduling. Please check logs for details", ex);
+                        "Internal Error - Exception thrown when scheduling. Please check logs for details", ex);
                 return;
             }
         }
         markFailedTopology(topologySubmitter, cluster, td, "Failed to schedule within " + maxSchedulingAttempts + " attempts");
     }
 
+    /*
+     * Class for tracking resources for scheduling a topology.
+     *
+     * Ideally we would simply track NormalizedResources, but shared topology memory complicates things.
+     * Topologies with shared memory may use more than the SharedMemoryLowerBound, and topologyRequiredResources
+     * ignores shared memory.
+     *
+     * Resources are tracked in two ways:
+     * 1) AvailableResources. Track cluster available resources and required topology resources.
+     * 2) RemainingRequiredResources. Start with required topology resources, and deduct for partially scheduled and evicted topologies.
+     */
+    private class TopologySchedulingResources {
+        ResourceMetrics metrics;
```

Comment:
It is not very clear to me why we need this

Reply:
`metrics` variable is used in NormalizedResourceOffer.remove(). It tracks when resources become negative. During scheduling, negative resources indicate a scheduling failure; otherwise resources should not become negative. I don't want to use Cluster metrics because (1) it causes testSchedulingNegativeResources to fail and (2) erroneous logging of (scheduling only) negative resources.

Maybe pass null metrics and change code to handle handle that?

Reply:
Yes I agree to change the code to handle it. 

We can have another function 
```
 public boolean remove(NormalizedResourcesWithMemory other) {
    return remove(other, null);
}
```
And then change `public boolean remove(NormalizedResourcesWithMemory other, ResourceMetrics resourceMetrics) ` to handle the case where `resourceMetrics` is null.

Reply:
removing `metrics` and making above changes